shepherd/Codegen.py

```diff
@@ -0,0 +1,368 @@
+import random
+
+class Codegen:
+
+    def __init__(self, rfids, rand_seed=None):
+        '''
+        Manages the generated code and rfid-answer pairs.
+        `rfids` must be an iterable of RFID ints.
+        Duplicate RFIDs in the provided iterable are ignored.
+        '''
+
+        # Seed our local RNG
+        self._random = random.Random(rand_seed)
+        rfids = set(rfids) # Ignore duplicates
+
+        # Keep generating random codes until we get one that gives a one-to-one
+        # mapping between RFIDs and answers
+        while True:
+            self._rfid_to_ans = {}
+            self._ans_to_rfid = {}
+            self._code = generate_challenge_code(self._random)
+            # We want there to be a one-to-one mapping between RFIDs and answers
+            # Therefore only break the loop if we randomly generate a code
+            # that produces unique answers for every RFID in the list
+            duplicate_answers = False
+            for rfid in rfids:
+                solution = staff_decode(self._code, rfid)
+                if solution in self._ans_to_rfid:
+                    duplicate_answers = True
+                    break
+                self._ans_to_rfid[solution] = rfid
+                self._rfid_to_ans[rfid] = solution
+            if not duplicate_answers:
+                break
+        assert len(self._rfid_to_ans) == len(self._ans_to_rfid)
+
+    def check_solution(self, solution):
+        '''
+        Returns which RFID that the solution corresponds to
+        return None if it doesn't match any.
+        '''
+        return self._ans_to_rfid.get(solution, None)
+
+    def get_code(self):
+        '''
+        Get the challenge code to be sent to students
+        '''
+        return self._code
+
+    def get_solution(self, rfid):
+        '''
+        Returns which solution that the RFID corresponds to
+        return None if it doesn't match any.
+
+        >>> c = Codegen([1, 2, 3], 0)
+        >>> c.check_solution(c.get_solution(1))
+        1
+        >>> c.check_solution(c.get_solution(2))
+        2
+        >>> c.check_solution(c.get_solution(3))
+        3
+        '''
+        return self._rfid_to_ans.get(rfid, None)
+
+# Various ideas for functions for the students
+# Included are example student implementations
+# Of course, the example solutions are optimized
+# for pedagogical simplicity, not runtime speed or memory
+
+# Functions presented in no particular order of difficulty
+# Both inputs and outputs guaranteed to be nonnegative integers
+
+def next_power(num):
```

Comment:
Are these functions going to be included in the Shepherd code base? I presumes yes, because it allows us to generate/check solutions. If not, these should be included as a separate file as "reference solutions". This perhaps could also just be precalculated and stored into a JSON or something, so that on the field these functions don't have to be run again (i.e. PSEUDOrandom). I think a next thing to consider is what the student skeleton code looks like.

Reply:
Actually all these functions are probably really fast to run, so don't need to precompute.